File: src/Sample/Plumix.Sample/Demos/General/DebugPaintingDemoPage.cs
using System;
using Avalonia;
using Avalonia.Media;
using Plumix.Rendering;
using Plumix.Widgets;

namespace Plumix;

// Dart parity source: dart_sample/lib/demos/general/debug_painting_demo_page.dart (exact sample parity)

public sealed class DebugPaintingDemoPage : StatefulWidget
{
    public override State CreateState() => new DebugPaintingDemoPageState();
}

internal sealed class DebugPaintingDemoPageState : State
{
    private bool _showPlaceholderChild;
    private bool _customGrid;

    public override Widget Build(BuildContext context)
    {
        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 12,
            children:
            [
                new Text("Placeholder + GridPaper", fontSize: 20, color: Colors.Black),
                new Text(
                    "Placeholder uses fallback dimensions only in unbounded space; GridPaper paints over its child.",
                    fontSize: 14,
                    color: Colors.DimGray),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildButton(
                            _showPlaceholderChild ? "Remove child" : "Add child",
                            () => SetState(() => _showPlaceholderChild = !_showPlaceholderChild)),
                        BuildButton(
                            _customGrid ? "Default grid" : "Custom grid",
                            () => SetState(() => _customGrid = !_customGrid)),
                    ]),
                new Expanded(
                    child: new Row(
                        crossAxisAlignment: CrossAxisAlignment.Stretch,
                        spacing: 16,
                        children:
                        [
                            new Expanded(child: BuildPlaceholderProbe()),
                            new Expanded(child: BuildGridPaperProbe()),
                        ])),
            ]);
    }

    private Widget BuildPlaceholderProbe()
    {
        return new Container(
            color: Colors.White,
            padding: new Thickness(12),
            child: new Column(
                crossAxisAlignment: CrossAxisAlignment.Stretch,
                spacing: 8,
                children:
                [
                    new Text("Unbounded fallback: 160 × 120", fontSize: 12, color: Colors.DarkSlateGray),
                    new Expanded(
                        child: new Align(
                            alignment: Alignment.TopLeft,
                            child: new UnconstrainedBox(
                                alignment: Alignment.TopLeft,
                                child: new Placeholder(
                                    color: Color.Parse("#FF455A64"),
                                    strokeWidth: 2,
                                    fallbackWidth: 160,
                                    fallbackHeight: 120,
                                    child: _showPlaceholderChild
                                        ? new Container(
                                            width: 96,
                                            height: 56,
                                            color: Color.Parse("#FFFFE8A3"),
                                            alignment: Alignment.Center,
                                            child: new Text("child", fontSize: 14, color: Colors.Black))
                                        : null)))),
                ]));
    }

    private Widget BuildGridPaperProbe()
    {
        return new Container(
            color: Colors.White,
            padding: new Thickness(12),
            child: new Column(
                crossAxisAlignment: CrossAxisAlignment.Stretch,
                spacing: 8,
                children:
                [
                    new Text(
                        _customGrid ? "interval=60, divisions=3, subdivisions=2" : "Flutter defaults",
                        fontSize: 12,
                        color: Colors.DarkSlateGray),
                    new Expanded(
                        child: new GridPaper(
                            color: _customGrid ? Color.Parse("#7FFF8A65") : null,
                            interval: _customGrid ? 60 : 100,
                            divisions: _customGrid ? 3 : 2,
                            subdivisions: _customGrid ? 2 : 5,
                            child: new Container(
                                color: Color.Parse("#FFF2F7FA"),
                                alignment: Alignment.Center,
                                child: new Text("foreground grid", fontSize: 14, color: Colors.Black)))),
                ]));
    }

    private static Widget BuildButton(string label, Action onTap)
    {
        return new SizedBox(
            width: 120,
            child: new CounterTapButton(
                label: label,
                onTap: onTap,
                background: Color.Parse("#FFDCE3ED"),
                foreground: Colors.Black,
                fontSize: 12,
                padding: new Thickness(10, 8)));
    }
}

File: dart_sample/lib/demos/general/debug_painting_demo_page.dart
import 'package:material_ui/material_ui.dart';

import '../../counter_widgets.dart';

class DebugPaintingDemoPage extends StatefulWidget {
  const DebugPaintingDemoPage({super.key});

  @override
  State<DebugPaintingDemoPage> createState() => _DebugPaintingDemoPageState();
}

class _DebugPaintingDemoPageState extends State<DebugPaintingDemoPage> {
  bool _showPlaceholderChild = false;
  bool _customGrid = false;
  bool _showErrors = false;

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 12,
      children: <Widget>[
        const Text(
          'Placeholder + GridPaper + ErrorWidget',
          style: TextStyle(fontSize: 20, color: Colors.black),
        ),
        const Text(
          'Placeholder uses fallback dimensions only in unbounded space; '
          'GridPaper paints over its child.',
          style: TextStyle(fontSize: 14, color: Colors.black54),
        ),
        Wrap(
          spacing: 8,
          runSpacing: 8,
          children: <Widget>[
            _buildButton(
              label: _showErrors ? 'Show grid' : 'Show errors',
              onTap: () {
                setState(() {
                  _showErrors = !_showErrors;
                });
              },
            ),
            _buildButton(
              label: _showPlaceholderChild ? 'Remove child' : 'Add child',
              onTap: () {
                setState(() {
                  _showPlaceholderChild = !_showPlaceholderChild;
                });
              },
            ),
            _buildButton(
              label: _customGrid ? 'Default grid' : 'Custom grid',
              onTap: () {
                setState(() {
                  _customGrid = !_customGrid;
                });
              },
            ),
          ],
        ),
        Expanded(
          child: _showErrors
              ? _buildErrorProbe()
              : Row(
                  crossAxisAlignment: CrossAxisAlignment.stretch,
                  spacing: 16,
                  children: <Widget>[
                    Expanded(child: _buildPlaceholderProbe()),
                    Expanded(child: _buildGridPaperProbe()),
                  ],
                ),
        ),
      ],
    );
  }

  static Widget _buildErrorProbe() {
    return Row(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 16,
      children: <Widget>[
        Expanded(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            spacing: 12,
            children: <Widget>[
              Expanded(
                child: ErrorWidget.withDetails(
                  message: 'Wide box: padding when space permits.',
                ),
              ),
              SizedBox(
                height: 100,
                child: ErrorWidget.withDetails(
                  message: 'Short box: no top padding.',
                ),
              ),
            ],
          ),
        ),
        SizedBox(
          width: 100,
          child: ErrorWidget.withDetails(
            message: 'Narrow box: no left padding.',
          ),
        ),
      ],
    );
  }

  Widget _buildPlaceholderProbe() {
    return Container(
      color: Colors.white,
      padding: const EdgeInsets.all(12),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        spacing: 8,
        children: <Widget>[
          const Text(
            'Unbounded fallback: 160 × 120',
            style: TextStyle(fontSize: 12, color: Colors.blueGrey),
          ),
          Expanded(
            child: Align(
              alignment: Alignment.topLeft,
              child: UnconstrainedBox(
                alignment: Alignment.topLeft,
                child: Placeholder(
                  color: const Color(0xFF455A64),
                  strokeWidth: 2,
                  fallbackWidth: 160,
                  fallbackHeight: 120,
                  child: _showPlaceholderChild
                      ? Container(
                          width: 96,
                          height: 56,
                          color: const Color(0xFFFFE8A3),
                          alignment: Alignment.center,
                          child: const Text(
                            'child',
                            style: TextStyle(fontSize: 14, color: Colors.black),
                          ),
                        )
                      : null,
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildGridPaperProbe() {
    return Container(
      color: Colors.white,
      padding: const EdgeInsets.all(12),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        spacing: 8,
        children: <Widget>[
          Text(
            _customGrid
                ? 'interval=60, divisions=3, subdivisions=2'
                : 'Flutter defaults',
            style: const TextStyle(fontSize: 12, color: Colors.blueGrey),
          ),
          Expanded(
            child: _customGrid
                ? GridPaper(
                    color: const Color(0x7FFF8A65),
                    interval: 60,
                    divisions: 3,
                    subdivisions: 2,
                    child: _buildGridChild(),
                  )
                : GridPaper(child: _buildGridChild()),
          ),
        ],
      ),
    );
  }

  static Widget _buildGridChild() {
    return Container(
      color: const Color(0xFFF2F7FA),
      alignment: Alignment.center,
      child: const Text(
        'foreground grid',
        style: TextStyle(fontSize: 14, color: Colors.black),
      ),
    );
  }

  static Widget _buildButton({
    required String label,
    required VoidCallback onTap,
  }) {
    return SizedBox(
      width: 120,
      child: CounterTapButton(
        label: label,
        onTap: onTap,
        background: const Color(0xFFDCE3ED),
        foreground: Colors.black,
        fontSize: 12,
        padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 8),
      ),
    );
  }
}
